### PR TITLE
Enable tp ops needed in ViT model

### PR DESCRIPTION
**Pull request by fduwjj**
This PR can be landed as prototyping or folks can continue working on this PR to upstream them. They are ops needed for Tensor Parallelism in ViT model.

### CODE REVIEW

File: test/spmd/tensor/test_matrix_ops.py [hunk 1]
```diff
@@ -1,31 +1,84 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
```
Comment (fduwjj) [inline test/spmd/tensor/test_matrix_ops.py:1]: Looks like the change comes with base change. Once the base got merged, I will rebase.

File: test/spmd/tensor/test_ddp.py [hunk 2]
```diff
@@ -96,7 +96,8 @@ def test_ddp_dist_tensor(self):
         sharded_input = Tensor.from_local(input, device_mesh, shard0_spec)
 
         # run DDP like a normal model
-        output = replicated_model(sharded_input)
```
Comment (fduwjj) [inline test/spmd/tensor/test_ddp.py:99]: TODO: Needs rebase.

File: spmd/tensor/ops/matrix_ops.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
```
Comment (fduwjj) [inline spmd/tensor/ops/matrix_ops.py:1]: TODO: Rebase and this is not part of this PR.

File: spmd/tensor/ops/matrix_ops.py [hunk 4]
```diff
@@ -16,50 +17,48 @@ def dist_addmm(
...
+    # TODO: implement all combinations
+
     local_input, local_mat1, local_mat2 = pytree.tree_map(
         unwrap_local_tensor, (input, mat1, mat2)
     )
     input_placement, mat1_placement, mat2_placement = pytree.tree_map(
         unwrap_single_placement, (input, mat1, mat2)
     )
     device_mesh = mat1.device_mesh
+    if (
```
Comment (wanchaol) [inline spmd/tensor/ops/matrix_ops.py:39]: maybe we should add the deleted cases as well for completeness.

File: spmd/tensor/ops/matrix_ops.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
 # implement matrix related ops for distributed tensor
+from locale import currency
```
Comment (wanchaol) [inline spmd/tensor/ops/matrix_ops.py:3]: nit: remove

File: spmd/tensor/api.py [hunk 6]
```diff
@@ -234,3 +275,8 @@ def device_mesh(self) -> DeviceMesh:
         # caller who want a different device_mesh
         # should call redistribute instead.
         return self._device_mesh
+
+    def contiguous(self) -> "Tensor":
```
Comment (wanchaol) [inline spmd/tensor/api.py:279]: can we add `_view_with_sharding_dim_change` and this one to the tp ops file as well? Let me add a `__torch_function__` method and you can register these two ops in the op table.

**Review by wanchaol — COMMENTED** [review]
some comments inlined, looks good overall. we might need to make the tests pass and linter happy?

File: spmd/tensor/api.py [hunk 7]
```diff
@@ -219,6 +222,44 @@ def redistribute(
...
 
+    def _view_with_sharding_dim_change(self, sharding_dim, shape):
+        if self.placements[0].is_shard():
+            return self.view(shape)
+        else:
+            if sharding_dim < 0:
+                sharding_dim += self.dim()
+
+            device_mesh = self.device_mesh
+            world_size = device_mesh._dim_groups[0].size()
```
Comment (wanchaol) [inline spmd/tensor/api.py:233]: nit: can we not using private pg variables to check the world_size? we have a explicit API for us to check the world size for the dim https://github.com/pytorch/PiPPy/blob/1f9ed8cfa93fcbb65d465912d12ad118f35fb86f/spmd/tensor/device_mesh.py#L206

File: spmd/tensor/ops/tp_sharding_ops.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    local_mat = pytree.tree_map(unwrap_local_tensor, self)
+    mat_placement = pytree.tree_map(unwrap_single_placement, self)
+
+    sharding_dim = mat_placement.dim
+    # When the sharding dim is negative, we need to ensure the new
+    # sharded tensor is still sharded by the original dimension.
+    if sharding_dim < 0:
+        sharding_dim = self.dim() + sharding_dim
+
+    world_size = self.device_mesh._dim_groups[0].size()
```
Comment (wanchaol) [inline spmd/tensor/ops/tp_sharding_ops.py:49]: same here

File: spmd/tensor/ops/tp_sharding_ops.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        local_tensor, tensor_list[0].device_mesh, tensor_list[0].placements
+    )
+
+
+@register_impl("aten.split.Tensor")
+def dist_split(self: Tensor, split_size_or_sections, dim=0) -> List[Tensor]:
+    local_mat = pytree.tree_map(unwrap_local_tensor, self)
+    mat_placement = pytree.tree_map(unwrap_single_placement, self)
+    sharding_dim = mat_placement.dim
+    world_size = self.device_mesh._dim_groups[0].size()
```
Comment (wanchaol) [inline spmd/tensor/ops/tp_sharding_ops.py:138]: ditto

File: spmd/tensor/api.py (hunk 7 above)
Comment (fduwjj) [reply spmd/tensor/api.py:233]: oh ok. Sure.

File: spmd/tensor/ops/matrix_ops.py (hunk 4 above)
Comment (fduwjj) [reply spmd/tensor/ops/matrix_ops.py:39]: Oh this should be part of Charlie's PR. Once this gets rebased, this part should be gone.

**Comment by fduwjj** [comment]
Commit has been messed a new clean PR is sent: https://github.com/pytorch/PiPPy/pull/327
